### PR TITLE
Add support for re-encryption termination for OpenShift routes.

### PR DESCRIPTION
**Pull request by f5gary**
Problem:
Add support for OpenShift reencrypt routes

Solution:
The main focus of this fix is the addition of reencrypt route support
in OpenShift, which required several other changes:
1. Added server SSL profile support to bigipconfigdriver.py and
   refactored existing SSL profile related code for reuse.
2. Added support for server SSL profiles in the controller, including
   the ability to attach them to virtual servers.
3. To improve configuration consistency, route objects are now sorted
   before they are processed by syncRoutes(). The sort order is by
   namespace/service/path.
4. Refactored createRSConfigFromRoute() to better handle the
   commonality between create and update modes while adding support for
   reencrypt routes.
5. OutputConfigLocked() now looks at server profiles when satisfying
   the iRule requirement for the passthrough/reencrypt iRule.
6. Corrected the handling of insecureEdgeTerminationPolicy and
   passthrough routes, as the policy only applies to edge terminated
   routes.
7. Added unit tests as appropriate.

affects-branches: master

### CODE REVIEW

File: pkg/appmanager/appManager_test.go [hunk 1]
```diff
@@ -3776,3 +3774,179 @@ func TestPassthroughRoute(t *testing.T) {
...
+	}
+	assert.True(haveClientSslProfile)
+	assert.True(haveServerSslProfile)
+
+	// and both should be referenced by the virtual
+	assert.Equal(1, rs.Virtual.GetProfileCountByContext(customProfileClient))
+	assert.Equal(1, rs.Virtual.GetProfileCountByContext(customProfileServer))
+}
+
+func TestRouteOrdering(t *testing.T) {
```
Comment (sjberman) [inline pkg/appmanager/appManager_test.go]: Maybe we move this into a routing_test.go file? Then we have a place to test other routing functions if we see fit.

File: pkg/appmanager/appManager.go [hunk 2]
```diff
@@ -161,6 +162,35 @@ func NewManager(params *Params) *Manager {
...
+		path := "/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt"
+		data, err := ioutil.ReadFile(path)
+		if nil != err {
+			log.Errorf("Unable to load default cluster certificate '%v': %v",
+				path, err)
+			return nil, false
+		}
+		appMgr.customProfiles.profs[skey] = NewCustomProfile(profile, string(data))
+	}
+	return &profile, !found
```
[the remark this answers is not in the document]
Comment (f5gary) [reply pkg/appmanager/appManager.go]: The boolean part of the return is if the cert was added and signals if there is a config change (i.e. the cert was not found), so the config needs to be written. I believe this is clearer from the perspective of the caller if the cert was added. I could add another local variable, set it to true when not found, and return it; but I'm not sure that helps.

File: pkg/appmanager/appManager.go [hunk 3]
```diff
@@ -1115,6 +1133,80 @@ func (appMgr *Manager) syncRoutes(
...
+
+func (appMgr *Manager) setServerSslProfile(
+	stats *vsSyncStats,
+	sKey serviceQueueKey,
+	rsCfg *ResourceConfig,
+	route *routeapi.Route,
+) {
+	if "" != route.Spec.TLS.DestinationCACertificate {
+		// Create new SSL server profile with the provided CA Certificate.
+		routeName := formatRouteRuleName(route)
```
[the remark this answers is not in the document]
Comment (f5gary) [reply pkg/appmanager/appManager.go]: Will do.

File: pkg/appmanager/appManager_test.go (hunk 1 above)
Comment (f5gary) [reply pkg/appmanager/appManager_test.go]: I will do this.

File: pkg/appmanager/appManager.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjberman) [reply pkg/appmanager/appManager.go]: Ok that makes sense.

**Pull request merged**